### PR TITLE
Rebased 2nd time full L1T O2O for 81X

### PR DESCRIPTION
**Pull request by kkotov**
This PR overrides https://github.com/cms-sw/cmssw/pull/15625

### CODE REVIEW

**Comment by franzoni** [comment]
please test

**Comment by mmusich** [comment]
@kkotov can you propagate https://github.com/kkotov/cmssw/commit/e774b1abc658e527231bd7746dc70ca63f900c47 and https://github.com/kkotov/cmssw/commit/b2fca995d095b31f7d91e9b04686b50244158aa4 here as well?

**Comment by kkotov** [comment]
@mmusich , indeed, the Clang error fixing commit (e774b1a) was in the original branch: https://github.com/cms-sw/cmssw/pull/15516/commits/9bb97ad2c7bc37524d4f8269adda71a1a6a064f3 (that is where I was cherry-picking it from), but somehow was lost while rebasing. The other commit (b2fca99) was not lost, but is titled differently: https://github.com/cms-sw/cmssw/pull/15842/commits/e385c959c41f0d851f4efc81fbe749eafe480108 . I checked the other commits and they all look fine. I'll reintroduce the first one again now ...

**Comment by kkotov** [comment]
@mmusich , ok, I've pushed the commit, but wait for me to test it myself first. I'll comment again here once I'm done. Thanks.

**Comment by franzoni** [comment]
please test

**Comment by kkotov** [comment]
@franzoni , I expect the test to fail this time again. it looks like I'll need to rebase yet another time. but let's see first...

**Comment by kkotov** [comment]
Indeed, after talking to Vladimir, I realized that there have been stuff that I depend, on merged yesterday. On top of that, Vladimir expects some more stuff to be merged today afternoon. Meaning that without the IB I'll need to fetch all this stuff in my local working directory, check if this works, and hope that it'll also work in the today's IB.

**Comment by rekovic** [comment]
@kkotov Indeed. After #15689 is merged into CMSSW_8_0_X branch later today, you will need to rebase on top of it.

**Comment by rekovic** [comment]
@kkotov  my above comment is relevant to your 80x PR #15846

Relevant to this 81x PR of yours, I will soon today make another 81x PR of L1Trigger code  (tag  l1t-tsg-v9.1 with NEW CALIBS). You might have to rebase once that is merged.

**Comment by rekovic** [comment]
@kkotov You should follow #15865 and its merging.  Check if you need to rebase after that.  My guess is probably not.

**Comment by rekovic** [comment]
+1

**Comment by mmusich** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
Hi @kkotov -while reviewing the rest
- please merge the L1*ParamsProducers packages into one. They are all very small
- these need to move to the usual data externals area
L1Trigger/L1TGlobal/data/Luminosity/startup/UGT_BASE_RS_ALGOBX_MASK_V1.xml
L1Trigger/L1TGlobal/data/Luminosity/startup/UGT_BASE_RS_FINOR_MASK_v17.xml
L1Trigger/L1TGlobal/data/Luminosity/startup/UGT_BASE_RS_PRESCALES_v10.xml
L1Trigger/L1TGlobal/data/Luminosity/startup/UGT_BASE_RS_VETO_MASK_v1.xml
(https://github.com/cms-data/L1Trigger-L1TMuon)

**Comment by kkotov** [comment]
thanks @rekovic , indeed the uGT files are now correctly merged to L1Trigger-L1TGlobal
@davidlange6 , rearrangement of packages under L1TriggerConfig is done

**Comment by kkotov** [comment]
Oh, the comment above should have gone to the https://github.com/cms-sw/cmssw/pull/15846 . Anyway, I guess we now have to do the same for this PR as well

**Comment by kkotov** [comment]
As I mentioned yesterday, it looks like I need to rebase. Nonetheless, the tests went fine. Let's try test the last change in this PR. If it fails, I'll PR another rebased branch: rebasedLastTimeFullL1TO2Ofor81X

**Comment by mulhearn** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by mmusich** [comment]
+1

**Comment by davidlange6** [comment]
@rekovic - could you also move the other luts (generally stuff in L1TMuon/data/) into the data area. All together they get big.

Thx-

File: CondFormats/L1TObjects/interface/L1MuDTEtaPatternLut.h [hunk 1]
```diff
@@ -73,7 +73,7 @@ class L1MuDTEtaPatternLut {
     /// return iterator which points to the one-past-last entry of the LUT
     inline ETFLut_iter end() const { return m_lut.end(); }
 
-  private:
+  public:
```
Comment (davidlange6) [inline CondFormats/L1TObjects/interface/L1MuDTEtaPatternLut.h]: @kkotov why this change is needed?

**Comment by rekovic** [comment]
@davidlange6   I thought I PRed all the data files from the O2O PR to externals:

https://github.com/cms-data/L1Trigger-L1TMuon/pull/7
https://github.com/cms-data/L1Trigger-L1TGlobal/pull/5

and I don't see them anymore in 

L1Trigger/L1TGlobal
L1Trigger/L1TMuon

in this PR here.  Are there any additional data files I should move?

**Comment by kkotov** [comment]
@davidlange6 , BMTF is using these ancient classes in their configuration, e.g.:
https://github.com/kkotov/cmssw/blob/rebased2ndTimeFullL1TO2Ofor81X/L1Trigger/L1TMuonBarrel/src/L1MuBMEtaProcessor.h#L136
On the other hand, Mike has moved the very same containers into the new BMTF CondFormats, e.g.:
https://github.com/kkotov/cmssw/blob/rebased2ndTimeFullL1TO2Ofor81X/CondFormats/L1TObjects/interface/L1TMuonBarrelParams.h#L76
As the result, I have to be able to copy the information between the two types of containers, as in:
https://github.com/kkotov/cmssw/blob/rebased2ndTimeFullL1TO2Ofor81X/L1Trigger/L1TMuonBarrel/src/L1TMuonBarrelParamsHelper.cc#L60-L61
Having the class representation public alleviates this problem great deal without a need to change the ancient code too much.

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
@rekovic right- but most of this pr is still data files.. Let's get them all moved into the more appropriate data area.

**Comment by kkotov** [comment]
@rekovic , @davidlange6 , David must be talking about files in https://github.com/kkotov/cmssw/tree/rebased2ndTimeFullL1TO2Ofor81X/L1Trigger/L1TMuon/data/ (there are some bmtf LUTs and few more xmls). I see these files are already available in the IBs, e.g.: /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_0_X_2016-09-18-1100/external/slc6_amd64_gcc530/data/L1Trigger/L1TMuon/data/ and /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-09-18-1100/external/slc6_amd64_gcc530/data/L1Trigger/L1TMuon/data/
I guess all it takes now is simply removing those from my PRs...

**Comment by rekovic** [comment]
please test

Now that data files are removed from the PR, restarting tests.

**Comment by rekovic** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by mmusich** [comment]
@davidlange6  @rekovic @kkotov was this PR reverted in 81x as well? 
Is there any progress in the understanding of the issue reported in the 80x version (#15846) ?

**Comment by rekovic** [comment]
@mmusich 

CMSSW 81X contains this PR without reverts.  

CMSSW 80x also contains the version of this PR (#15846). That PR was briefly reverted and then soon after re-reverted, so it is in also.

**Comment by davidlange6** [comment]
Hi @musich - I believe this issue is resolved in both 80x and 81x by @rekovic. Do you see otherwise?

**Comment by rekovic** [comment]
@mmusich  Unless you are talking about reverting of the PR, can you say more precisely which of issues you are asking about?

**Comment by mmusich** [comment]
I am admittedly lost. Where is the follow-up PR on https://github.com/cms-sw/cmssw/pull/15846?

**Comment by koskot77** [comment]
@mmusich , @rekovic , @davidlange6 ,
I think there was some confusion triggered by first bullet on slide #4 of Giacomo's talk:
https://indico.cern.ch/event/565733/contributions/2285157/attachments/1347281/2032038/2016-10-03_AlCaDB_news.pdf
Btw, @mmusich I often cannot connect to the AlCa meeting because of the Vidyo PIN. Could you remind me what the PIN is or where it is announced? Thanks.

**Comment by mmusich** [comment]
@kkotov hope this helps:

screenshot 2016-10-03 16 53 23

**Comment by koskot77** [comment]
@mmusich oh, I never noticed the arrow in the top right corner before. Thanks!

**Comment by franzoni** [comment]
hello, @rekovic , @davidlange6 , @kkotov 
we had lost track of the re-re-verts. Thanks to @kkotov for clarifying verbally.
We've updated the slides, see: https://indico.cern.ch/event/565733/contributions/2285157/

**Comment by rekovic** [comment]
@franzoni Thank you.  

Sorry for not being more explicit with the information on this github discussion here.  Here is some historical information.

Full O2O PR was finally integrated (AKA revert of reverted merge) into the CMSSW_8_0_X release 10 days ago.  The CMSSW_8_0_X_2016-09-22-2300  IB already contained it.

<pre>
* | | | | | | |   6c07c74 (tag: CMSSW_8_0_X_2016-09-23-1100, tag: CMSSW_8_0_X_2016-09-22-2300, tag: CMSSW_8_0_THREADED_X_2016-09-23-1100, tag: CMSSW_8_0_THREADED_X_2016-09-22-2300) Merge pull request #159
|\ \ \ \ \ \ \ \  
| * | | | | | | | 52c8737 (official-cmssw/revert-15947-revert-15846-rebasedFullL1TO2Ofor80X, cms-sw/revert-15947-revert-15846-rebasedFullL1TO2Ofor80X) Revert "Revert "Rebased full L1T O2O for 80X""
</pre>
